### DISCUSSION TITLE
Issues with Async Lazy Loading Nested Relationships in SQLAlchemy

### DISCUSSION THREAD

**Discussion by jtfidje** [sqlalchemy/sqlalchemy#11258] [Usage Questions] [answered]
'm currently working with SQLAlchemy's async session and I'm facing an issue with nested relationships and lazy loading. I've tried my best searching for a solution, but haven't found one yet. I posted a question on Gitter and was encouraged to try my luck here 🙂 

In my setup, I have two models, User and Organization, with a many-to-many relationship through an association table user_organization_association.

Here's a simplified version of my models:

```python
class Base(AsyncAttrs, DeclarativeBase):
    ...

user_organization_association = Table(
    "user_organization_association",
    Base.metadata,
    Column("user_id", ForeignKey("user.id")),
    Column("organization_id", ForeignKey("organization.id")),
)

class Organization(Base):
    __tablename__ = "organization"
    # ... (Other fields)
    users: Mapped[list["User"]] = relationship(
        secondary=user_organization_association,
        lazy="selectin",
        back_populates="organizations",
    )

class User(Base):
    __tablename__ = "user"
    # ... (Other fields)
    organizations: Mapped[list[Organization]] = relationship(
        secondary=user_organization_association, lazy="selectin", back_populates="users"
    )
```

When I access User.organizations[0], it loads as expected. However, when I try to access User.organizations[0].users[0], it does not trigger a lazy load and instead raises a MissingGreenlet: greenlet_spawn has not been called exception.

My goal is to achieve the following behavior:

- When accessing User.organizations, I want to load a list of full Organization objects.
- When accessing Organization.users, I want to load a list of full User objects.
- However, for nested relationships, I want to load only the IDs to avoid deep recursive loading. That is, User.organizations[...].users should be a list of User.id, and Organization.users[...].organizations should be a list of Organization.id.

I'm struggling to configure the relationships to work as described above. Is there a recommended pattern for this use case?

Please let me know if you need any more information. Any help or pointers would be greatly appreciated!

P.S.: Here is a Pastebin with code that hopefully will help you reproduce the error I get: https://pastebin.com/ySaGPz29
Running the code in the above pastebin succeeds the first time, but fails all runs after that.

**Comment by CaselIT** [comment, marked as answer]
Hi,

This behaviour is controlled by `join_depth` https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.join_depth

setting it to 2 on users makes the code work.

**Reply by jtfidje** [reply]
Aah thank you - I completely missed this in the docs.

**Reply by CaselIT** [reply]
that's probably not a great name.

@zzzeek do you thing renaming this would create more issues? Maybe just better docs could be appropriate? Like linking since from the selectinload docs?
